File: Tree Algorithms/Counting_Paths.cpp
// Choose cpp when the time
// limit will for sure give tle in pypy
#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
using namespace __gnu_pbds;
using namespace std;

#define int long long
#define INF LLONG_MAX
#define fastio ios::sync_with_stdio(false); cin.tie(0);
#define len(arr) arr.size()
#define f first
#define s second
#define pb push_back
#define all(x) x.begin(), x.end()
#define range(i, n) for (int i = 0; i < (n); ++i)
#define rangea(i, a, b) for (int i = (a); i < (b); ++i)
#define MOD1 998244353
#define MOD 1000000007
using vi = vector<int>;
using vii = vector<vector<int>>;
using pi = pair<int,int>;
#pragma GCC optimize("O3,unroll-loops")
#pragma GCC target("avx2,bmi,bmi2,lzcnt,popcnt")

template <typename T>
using ordered_multiset = tree<
    std::pair<T, int>,
    null_type,
    std::less<std::pair<T, int>>,
    rb_tree_tag,
    tree_order_statistics_node_update
>;
template<typename T> istream& operator>>(istream& in, vector<T>& v) {for (auto& x : v) in >> x;return in;}
template<typename T1, typename T2> istream& operator>>(istream& in, pair<T1, T2>& p) {in >> p.first >> p.second;return in;}
template<typename T>
class is_printable {
    template<typename U> static auto test(int) -> decltype(cout << declval<U>(), true_type{});
    template<typename> static auto test(...) -> false_type;
public: static constexpr bool value = decltype(test<T>(0))::value;
};
template<typename T> enable_if_t<is_printable<T>::value> __print(const T& val) { cout << val; }
template<typename T1, typename T2> void __print(const pair<T1, T2>& p) {
    cout << "("; __print(p.first); cout << ":"; __print(p.second); cout << ")";
}
template<typename T> enable_if_t<!is_printable<T>::value> __print(const T& container) {
    cout << "{"; bool first = true;
    for (const auto& x : container) { if (!first) cout << ", "; __print(x); first = false; }
    cout << "}";
}
void print() { cout << "\n"; }
template<typename T, typename... Args>
void print(const T& t, const Args&... rest) {
    __print(t); if constexpr (sizeof...(rest)) cout << " "; print(rest...);
}

int gcd(int a, int b) {if (b == 0) return a;return gcd(b, a % b);}
int lcm(int a, int b) {return a / gcd(a, b) * b;}
int II(){int a;cin>>a;return a;}
string SI(){string s;cin>>s;return s;}
vi LII(int n){vi a(n);cin>>a;return a;}

// String hashing: sh/shclass, Number: numtheory, SparseTable: SparseTable, SortedList: sortl
// Segment Tree(lazy propogation): SegmentTree, Merge Sort Tree: sorttree, Trie/Treap: Tries
// binary indexed tree: BIT, Segment Tree(point updates): SegmentPoint, Convex Hull: hull
// Combinatorics: pnc, Diophantine Equations: dpheq, Graphs: graphs, Centroid Decomp.: graph_decom
// Persistent Segment Tree: perseg, FreqGraphs: bgraph, GrapthOth: graphoth, DSU: DSU, FFT:fft
// Template : https://github.com/OmAmar106/Template-for-Competetive-Programming

vector<int> bellman_ford(int n, vector<tuple<int, int, int>> &edges, int start){
    vector<int> dist(n, INF), pred(n, -1);
    dist[start] = 0;
    for (int i = 0; i < n - 1; i++) {
        for (int i = 0; i < edges.size(); i++) {
        int u = get<0>(edges[i]);
        int v = get<1>(edges[i]);
        int d = get<2>(edges[i]);
            if (dist[u] != INF && dist[u] + d < dist[v]) {
                dist[v] = dist[u] + d;
                pred[v] = u;
            }
        }
    }
    return dist;
}

class BinaryLift {
public:
    int n, L;
    vector<int> depth;
    vector<vector<int>> parent;

    BinaryLift(vector<vector<int>>& graph, int root = 0) {
        n = graph.size();
        L = 32 - __builtin_clz(n);
        parent.assign(L, vector<int>(n, -1));
        depth.assign(n, -1);
        queue<int> bfs;
        bfs.push(root);
        depth[root] = 0;
        while (!bfs.empty()) {
            int node = bfs.front(); bfs.pop();
            for (int nei : graph[node]) {
                if (depth[nei] == -1) {
                    parent[0][nei] = node;
                    depth[nei] = depth[node] + 1;
                    bfs.push(nei);
                }
            }
        }
        for (int i = 1; i < L; i++) {
            for (int v = 0; v < n; v++) {
                if (parent[i - 1][v] != -1)
                    parent[i][v] = parent[i - 1][parent[i - 1][v]];
            }
        }
    }

    int lca(int a, int b) {
        if (depth[a] < depth[b]) swap(a, b);
        int d = depth[a] - depth[b];
        for (int i = 0; i < L; i++) {
            if (d & (1LL << i)) a = parent[i][a];
        }
        if (a == b) return a;
        for (int i = L - 1; i >= 0; i--) {
            if (parent[i][a] != parent[i][b]) {
                a = parent[i][a];
                b = parent[i][b];
            }
        }
        return parent[0][a];
    }

    int distance(int a, int b) {
        return depth[a] + depth[b] - 2 * depth[lca(a, b)];
    }

    int kth_ancestor(int a, int k) {
        if (depth[a] < k) return -1;
        for (int i = 0; i < L; i++) {
            if (k & (1LL << i)) a = parent[i][a];
        }
        return a;
    }
};

vector<int> kahn(vector<vector<int>>& graph) {
    int n = graph.size();
    vector<int> indeg(n, 0), res;
    for (int i = 0; i < n; i++)
        for (int e : graph[i])
            indeg[e]++;
    queue<int> q;
    for (int i = 0; i < n; i++)
        if (indeg[i] == 0) q.push(i);
    while (!q.empty()) {
        int node = q.front(); q.pop();
        res.push_back(node);
        for (int e : graph[node]) {
            indeg[e]--;
            if (indeg[e] == 0) q.push(e);
        }
    }
    return res.size() == n ? res : vector<int>();
}

vector<pair<int, int>> dfs(vector<vector<int>>& graph) {
    int n = graph.size(), time = 0;
    vector<pair<int, int>> starttime(n, {0, 0});
    stack<tuple<int, int, int>> stack;
    stack.emplace(0, -1, 0);
    while (!stack.empty()) {
        int cur, prev, state;
        cur = get<0>(stack.top());
        prev = get<1>(stack.top());
        state = get<2>(stack.top());
        stack.pop();
        if (state == 0) {
            starttime[cur].first = time++;
            stack.emplace(cur, prev, 1);
            for (int neighbor : graph[cur]) {
                if (neighbor == prev) continue;
                stack.emplace(neighbor, cur, 0);
            }
        } else {
            starttime[cur].second = time;
        }
    }
    return starttime;
}

vector<int> euler_path(unordered_map<int, multiset<int>>& d) {
    vector<int> ans;
    stack<int> start;
    start.push(1);
    while (!start.empty()) {
        int cur = start.top();
        if (d[cur].empty()) {
            ans.push_back(cur);
            start.pop();
        } else {
            int k1 = *d[cur].begin();
            d[cur].erase(d[cur].begin());
            d[k1].erase(d[k1].find(cur));
            start.push(k1);
        }
    }
    return ans;
}

class TwoSat {
public:
    int n;
    vector<vector<int>> graph;

    TwoSat(int n) : n(n), graph(2 * n) {}

    int negate(int x) {
        return x < n ? x + n : x - n;
    }

    void _imply(int x, int y) {
        graph[x].push_back(y);
        graph[negate(y)].push_back(negate(x));
    }

    void either(int x, int y) {
        _imply(negate(x), y);
        _imply(negate(y), x);
    }

    void set(int x) {
        _imply(negate(x), x);
    }

    vector<vector<int>> find_SCC() {
        int n = graph.size();
        vector<int> order, comp(n, -1), low(n), depth(n, -1);
        vector<vector<int>> SCC;
        stack<int> st;
        int timer = 0;

        function<void(int)> dfs = [&](int v) {
            low[v] = depth[v] = timer++;
            st.push(v);
            for (int u : graph[v]) {
                if (depth[u] == -1) dfs(u);
                if (comp[u] == -1) low[v] = min(low[v], low[u]);
            }
            if (low[v] == depth[v]) {
                SCC.emplace_back();
                while (true) {
                    int u = st.top(); st.pop();
                    comp[u] = SCC.size() - 1;
                    SCC.back().push_back(u);
                    if (u == v) break;
                }
            }
        };

        for (int i = 0; i < n; i++) {
            if (depth[i] == -1) dfs(i);
        }
        return SCC;
    }

    pair<bool, vector<int>> solve() {
        auto SCC = find_SCC();
        vector<int> order(2 * n, 0);
        for (int i = 0; i < SCC.size(); i++)
            for (int x : SCC[i])
                order[x] = i;
        vector<int> res(n);
        for (int i = 0; i < n; i++) {
            if (order[i] == order[negate(i)])
                return {false, {}};
            res[i] = order[i] > order[negate(i)];
        }
        return {true, res};
    }
};

void solve() {
    int n,m;
    cin>>n>>m;
    int u,v;
    vii d(n);
    for(int i=0;i<n-1;i++){
        cin>>u>>v;
        u--;v--;
        d[u].pb(v);
        d[v].pb(u);
    }

    vi ans(n,0);

    BinaryLift bl = BinaryLift(d);
    vi par = bl.parent[0];

    while(m--){
        cin>>u>>v;
        u--;v--;
        ans[u]++;
        ans[v]++;
        int f = bl.lca(u,v);
        ans[f]--;
        if(f){
            ans[par[f]]--;
        }
    }

    vector<bool> visited(n,false);

    auto dfs = [&](auto&& self,int cur,int prev,vii &d,vi &ans) -> int {
        int val = 0;
        for(auto it:d[cur]){
            if(it!=prev){
                val += self(self,it,cur,d,ans);
            }
        }
        ans[cur] += val;
        return ans[cur];
    };

    dfs(dfs,0,-1,d,ans);
    
    for(auto it:ans){
        cout<<it<<" ";
    }

}

int32_t main() {
    fastio
    int t=1;
    while (t--) solve();
    return 0;
}

File: Tree Algorithms/Counting_Paths.py
import sys,math,cmath,random,os
from heapq import heappush,heappop
from bisect import bisect_right,bisect_left
from collections import Counter,deque,defaultdict
from itertools import permutations,combinations
from io import BytesIO, IOBase
from decimal import Decimal,getcontext
 
BUFSIZE = 8192
class FastIO(IOBase):
    newlines = 0
    def __init__(self, file):
        self._file = file
        self._fd = file.fileno()
        self.buffer = BytesIO()
        self.writable = "x" in file.mode or "r" not in file.mode
        self.write = self.buffer.write if self.writable else None
    def read(self):
        while True:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            if not b:
                break
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines = 0
        return self.buffer.read()
    def readline(self):
        while self.newlines == 0:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            self.newlines = b.count(b"\n") + (not b)
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines -= 1
        return self.buffer.readline()
    def flush(self):
        if self.writable:
            os.write(self._fd, self.buffer.getvalue())
            self.buffer.truncate(0), self.buffer.seek(0)
class IOWrapper(IOBase):
    def __init__(self, file):
        self.buffer = FastIO(file)
        self.flush = self.buffer.flush
        self.writable = self.buffer.writable
        self.write = lambda s: self.buffer.write(s.encode("ascii"))
        self.read = lambda: self.buffer.read().decode("ascii")
        self.readline = lambda: self.buffer.readline().decode("ascii")
sys.stdin, sys.stdout = IOWrapper(sys.stdin), IOWrapper(sys.stdout)
 
# functions #
# MOD = 998244353
MOD = 10**9 + 7
RANDOM = random.randrange(1,2**62)
def gcd(a,b):
    while b:
        a,b = b,a%b
    return a
def lcm(a,b):
    return a//gcd(a,b)*b
def w(x):
    return x ^ RANDOM
II = lambda : int(sys.stdin.readline().strip())
LII = lambda : list(map(int, sys.stdin.readline().split()))
MI = lambda x : x(map(int, sys.stdin.readline().split()))
SI = lambda : sys.stdin.readline().strip()
SLI = lambda : list(map(lambda x:ord(x)-97,sys.stdin.readline().strip()))
LII_1 = lambda : list(map(lambda x:int(x)-1, sys.stdin.readline().split()))
LII_C = lambda x : list(map(x, sys.stdin.readline().split()))
MATI = lambda x : [list(map(int, sys.stdin.readline().split())) for _ in range(x)]
##
 
#String hashing: shclass, fenwick sortedlist: fsortl, Number: numtheory/numrare, SparseTable: SparseTable
#Bucket Sorted list: bsortl, Segment Tree(lp,selfop): SegmentTree, bootstrap: bootstrap, Trie: tries
#binary indexed tree: BIT, Segment Tree(point updates): SegmentPoint, Convex Hull: hull, BitArray: bitarray
#Combinatorics: pnc, Diophantine Equations: dpheq, DSU: DSU, Geometry: Geometry, FFT: fft, XOR_dict: xdict
#Persistent Segment Tree: perseg, Binary Trie: b_trie, HLD: hld, String funcs: sf, Segment Tree(lp): SegmentOther
#Graph1(dnc,bl): graphadv, Graph2(khn,sat): 2sat, Graph3(fltn,bprt): graphflatten, Graph4(ep,tp,fw,bmf): graphoth
#Graph5(djik,bfs,dfs): graph, Graph6(dfsin): dfsin, utils: utils, Persistent DSU: perdsu, Merge Sort Tree: sorttree
#2-D BIT: 2DBIT, MonoDeque: mono, nummat: matrix, SuffixAutomaton: sautomaton, linalg: linalg, SquareRtDecomp: sqrt
#Grapth7(bridges): graph_dmgt, FWHT(^,|,&): fwht, Graph8(centr_decom): graph_decom, DpOptimize(knth,dnc): dpopt
#Template : https://github.com/OmAmar106/Template-for-Competetive-Programming
#if os.environ.get('LOCAL'):sys.stdin = open(r'input.txt', 'r');sys.stdout = sys.stderr = open(r'output.txt','w')
#if os.environ.get('LOCAL'):import hashlib;print('Hash Value :',hashlib.md5(open(__file__, 'rb').read()).hexdigest());
 
def extras():
    getcontext().prec = 50
    sys.setrecursionlimit(10**6)
    sys.set_int_max_str_digits(10**5)
# extras()
 
def interactive():
    import builtins
    # print(globals())
    globals()['print'] = lambda *args, **kwargs: builtins.print(*args, flush=True, **kwargs)
# interactive()
 
def GI(n,m=None,sub=-1,dirs=False,weight=False):
    if m==None:
        m = n-1
    d = [[] for i in range(n)]
    if not weight:
        for i in range(m):
            u,v = LII_C(lambda x:int(x)+sub)
            d[u].append(v)
            if not dirs:
                d[v].append(u)
    else:
        for i in range(m):
            u,v,w = LII()
            d[u+sub].append((v+sub,w))
            if not dirs:
                d[v+sub].append((u+sub,w))
    return d
 
ordalp = lambda s : ord(s)-65 if s.isupper() else ord(s)-97
alp = lambda x : chr(97+x)
yes = lambda : print("Yes")
no = lambda : print("No")
yn = lambda flag : print("Yes" if flag else "No")
printf = lambda x : print(-1 if x==float('inf') else x)
lalp = 'abcdefghijklmnopqrstuvwxyz'
ualp = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
dirs = ((1,0),(0,1),(-1,0),(0,-1))
dirs8 = ((1,0),(0,1),(-1,0),(0,-1),(1,-1),(-1,1),(1,1),(-1,-1))
ldir = {'D':(1,0),'U':(-1,0),'R':(0,1),'L':(0,-1)}
 
INF = float("inf")
class Dinic:
    def __init__(self, n):
        self.lvl = [0] * n
        self.ptr = [0] * n
        self.q = [0] * n
        self.adj = [[] for _ in range(n)]
    def add_edge(self, a, b, c, rcap=0):
        self.adj[a].append([b, len(self.adj[b]), c, 0])
        self.adj[b].append([a, len(self.adj[a]) - 1, rcap, 0])
    def dfs(self, v, t, f):
        if v == t or not f:
            return f
        for i in range(self.ptr[v], len(self.adj[v])):
            e = self.adj[v][i]
            if self.lvl[e[0]] == self.lvl[v] + 1:
                p = self.dfs(e[0], t, min(f, e[2] - e[3]))
                if p:
                    self.adj[v][i][3] += p
                    self.adj[e[0]][e[1]][3] -= p
                    return p
            self.ptr[v] += 1
        return 0
    def calc(self, s, t):
        flow, self.q[0] = 0, s
        for l in range(31):  # l = 30 maybe faster for random data
            while True:
                self.lvl, self.ptr = [0] * len(self.q), [0] * len(self.q)
                qi, qe, self.lvl[s] = 0, 1, 1
                while qi < qe and not self.lvl[t]:
                    v = self.q[qi]
                    qi += 1
                    for e in self.adj[v]:
                        if not self.lvl[e[0]] and (e[2] - e[3]) >> (30 - l):
                            self.q[qe] = e[0]
                            qe += 1
                            self.lvl[e[0]] = self.lvl[v] + 1
                p = self.dfs(s, t, INF)
                while p:
                    flow += p
                    p = self.dfs(s, t, INF)
                if not self.lvl[t]:
                    break
        return flow
 
class AuxiliaryTree:
    def __init__(self, edge, root = 0):
        self.n = len(edge)
        self.order = [-1] * self.n
        self.path = [-1] * (self.n-1)
        self.depth = [0] * self.n
        if self.n == 1: return
        parent = [-1] * self.n
        que = [root]
        t = -1
        while que:
            u = que.pop()
            self.path[t] = parent[u]
            t += 1
            self.order[u] = t
            for v in edge[u]:
                if self.order[v] == -1:
                    que.append(v)
                    parent[v] = u
                    self.depth[v] = self.depth[u] + 1
        self.n -= 1
        self.h = self.n.bit_length()
        self.data = [0] * (self.n * self.h)
        self.data[:self.n] = [self.order[u] for u in self.path]
        for i in range(1, self.h):
            for j in range(self.n - (1<<i) + 1):
                self.data[i*self.n + j] = min(self.data[(i-1)*self.n + j], self.data[(i-1)*self.n + j+(1<<(i-1))])
 
    def lca(self, u, v):
        if u == v: return u
        l = self.order[u]
        r = self.order[v]
        if l > r:
            l,r = r,l
        level = (r - l).bit_length() - 1
        return self.path[min(self.data[level*self.n + l], self.data[level*self.n + r-(1<<level)])]
 
    def dis(self, u, v):
        if u == v: return 0
        l = self.order[u]
        r = self.order[v]
        if l > r:
            l,r = r,l
        level = (r - l).bit_length() - 1
        p = self.path[min(self.data[level*self.n + l], self.data[level*self.n + r-(1<<level)])]
        return self.depth[u] + self.depth[v] - 2 * self.depth[p]
 
    def make(self, vs):
        k = len(vs)
        vs.sort(key = self.order.__getitem__)
 
        par = dict()
        edge = dict()
        edge[vs[0]] = []
 
        st = [vs[0]]
 
        for i in range(k - 1):
            l = self.order[vs[i]]
            r = self.order[vs[i+1]]
            level = (r - l).bit_length() - 1
            w = self.path[min(self.data[level*self.n + l], self.data[level*self.n + r-(1<<level)])]
            if w != vs[i]:
                p = st.pop()
                while st and self.depth[w] < self.depth[st[-1]]:
                    par[p] = st[-1]
                    edge[st[-1]].append(p)
                    p = st.pop()
 
                if not st or st[-1] != w:
                    st.append(w)
                    edge[w] = [p]
                else:
                    edge[w].append(p)
                par[p] = w
 
            st.append(vs[i+1])
            edge[vs[i+1]] = []
 
        for i in range(len(st) - 1):
            edge[st[i]].append(st[i+1])
            par[st[i+1]] = st[i]
 
        par[st[0]] = -1
        return st[0], edge, par
 
class binary_lift:
    def __init__(self, graph, f=max, root=0, flag=False):
        n = len(graph)
        parent = [-1] * (n + 1)
        depth = self.depth = [-1] * n
        bfs = [root]
        depth[root] = 0
        # data = [0]*n
        for node in bfs:
            # for nei,w in graph[node]:
            for nei in graph[node]:
                if depth[nei] == -1:
                    # data[nei] = w
                    parent[nei] = node
                    depth[nei] = depth[node] + 1
                    bfs.append(nei)
        parent = self.parent = [parent]
        self.f = f
        if flag:
            data = self.data = [data]
            for _ in range(max(depth).bit_length()):
                old_data = data[-1]
                old_parent = parent[-1]
                data.append([f(val, old_data[p]) for val,p in zip(old_data, old_parent)])
                parent.append([old_parent[p] for p in old_parent])
        else:
            for _ in range(max(depth).bit_length()):
                old_parent = parent[-1]
                parent.append([old_parent[p] for p in old_parent])
    def lca(self, a, b):
        depth = self.depth
        parent = self.parent
        if depth[a] < depth[b]:
            a,b = b,a
        d = depth[a] - depth[b]
        for i in range(d.bit_length()):
            if (d >> i) & 1:
                a = parent[i][a]
        for i in range(depth[a].bit_length())[::-1]:
            if parent[i][a] != parent[i][b]:
                a = parent[i][a]
                b = parent[i][b]
        if a != b:
            return parent[0][a]
        else:
            return a
    def distance(self, a, b):
        return self.depth[a] + self.depth[b] - 2 * self.depth[self.lca(a,b)]
    def kth_ancestor(self, a, k):
        parent = self.parent
        if self.depth[a] < k:
            return -1
        for i in range(k.bit_length()):
            if (k >> i) & 1:
                a = parent[i][a]
        return a
    def __call__(self, a, b, c=0):
        depth = self.depth
        parent = self.parent
        data = self.data
        f = self.f
        c = self.lca(a, b)
        val = c
        for x,d in (a, depth[a] - depth[c]), (b, depth[b] - depth[c]):
            for i in range(d.bit_length()):
                if (d >> i) & 1:
                    val = f(val, data[i][x])
                    x = parent[i][x]
        return val
 
def solve():
    n,m = LII()
    d = GI(n)
 
    ans = [0]*n
 
    bl = binary_lift(d)
    par = bl.parent[0]
 
    for _ in range(m):
        u,v = LII_1()
        f = bl.lca(u,v)
        ans[u] += 1
        ans[v] += 1
        ans[f] -= 1
        if f:
            ans[par[f]] -= 1
 
    visited = [False]*n
 
    st = [0]
    while st:
        start = st[-1]
        if not visited[start]:
            visited[start] = True
            for j in d[start]:
                if not visited[j]:
                    st.append(j)
        else:
            st.pop()
            for j in d[start]:
                if j!=par[start]:
                    ans[start] += ans[j]
    
    print(*ans)
 
    #L1 = LII()
    #st = SI()
solve()